internal/lockclient/simple_client.go

```diff
@@ -20,114 +26,306 @@ type SimpleClient struct {
 	config *lockservice.SimpleConfig
 	cache  *cache.LRUCache
 	mu     sync.Mutex
+	id     id.ID
+	log    zerolog.Logger
+
+	// sessions holds the mapping of a process to a session.
+	sessions map[id.ID]session.Session
+	// sessionTimers maintains the timers for each session,
+	sessionTimers map[id.ID]chan struct{}
+	// sessionAcquisitions has a list of all the acquisitions
+	// from a particular process. This has no knowledge of
+	// whether the process owning the lock has an active session
+	// or not, this guarantee has to be ensured by the client.
+	sessionAcquisitions map[id.ID][]lockservice.Descriptors
 }
 
 // NewSimpleClient returns a new SimpleClient of the given parameters.
 // This client works with or without the existance of a cache.
-func NewSimpleClient(config *lockservice.SimpleConfig, cache *cache.LRUCache) *SimpleClient {
+func NewSimpleClient(config *lockservice.SimpleConfig, log zerolog.Logger, cache *cache.LRUCache) *SimpleClient {
+	clientID := id.Create()
+	sessions := make(map[id.ID]session.Session)
+	sessionTimers := make(map[id.ID]chan struct{})
+	sessionAcquisitions := make(map[id.ID][]lockservice.Descriptors)
 	return &SimpleClient{
-		config: config,
-		cache:  cache,
+		config:              config,
+		cache:               cache,
+		id:                  clientID,
+		log:                 log,
+		sessions:            sessions,
+		sessionTimers:       sessionTimers,
+		sessionAcquisitions: sessionAcquisitions,
 	}
 }
 
 var _ Client = (*SimpleClient)(nil)
 
+// Connect lets the user process to establish a connection with the
+// client.
+func (sc *SimpleClient) Connect() session.Session {
+	sessionID := id.Create()
+	processID := id.Create()
+	session := session.NewSession(sessionID, sc.id, processID)
+	sc.sessions[processID] = session
+	sc.startSession(processID)
+	sc.log.
+		Debug().
+		Str(processID.String(), "connected").
+		Msg("session created")
+	return session
+}
+
 // Acquire allows the user process to acquire a lock.
-func (sc *SimpleClient) Acquire(d lockservice.Descriptors) error {
-	return sc.acquire(d)
+// This returns a "session expired" error if the session expires when
+// the lock is being acquired.
+//
+// All locks acquired during the session will be revoked if the session
+// expires.
+func (sc *SimpleClient) Acquire(d lockservice.Object, s session.Session) error {
+	sc.mu.Lock()
+	if _, ok := sc.sessions[s.ProcessID()]; !ok {
+		sc.mu.Unlock()
+		return ErrSessionInexistent
+	}
+	sc.mu.Unlock()
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	close := make(chan struct{}, 1)
+	go func() {
+		for {
+			sc.mu.Lock()
```

Comment:
I think instead of having a single mutex for the whole client, it would make sense to have a mutex per process. because if MANY clients are connected to the client, and all of the call acquire() at around the same time then they'll all just be waiting (because others might get the lock first) but would ideally like for them to acquire there respective locks right?

Reply:
I like the way you've done this though! really nice :)

Reply:
You really missed the point of locks in that question. Locks are meant to serialise things. As you can see, we have a COMMON MAP `session*`, which when used in multiple goroutines cause a race. Thus a lock and thus a single lock.